Laser phase measurement (wavefront-sensor) and non-Gaussian intensity distribution in the focal spot

Hi everyone,

It is great to know (from slack discussion) that LASY is soon going to be implemented for use in FBPIC. I am very excited about this additional powerful utility.

**During the slack discussion with Remi, I mis-understood 'spectral phase' with the 'phase measured with the wave-front sensor'.** 

Non-uniform 'spectral-phase' tells that the laser pulse is not compressed to Fourier limit, however, it can not give any information about laser spot (please correct me if I am wrong). It is nicely available for use in the code (Paraxial_Approximation_Laser).

On the other hand, the non-uniform phase distribution, measured by wave-front sensor, tells how much the focal spot is deviated from Gaussian one (please correct me if I am wrong). 

**Do you think that we can load the phase distribution information measured by wave-front sensor in code and calculate the real intensity distribution in the laser spot. If this can be done, it will be extremely powerful and realistic.** 

Thank you and my best wishes to the entire team.

Hi @MaxThevenet ,

I used the example = https://github.com/LASY-org/lasy/blob/development/examples/example_modal_decomposition_data.py

**and got the following error**

C:\Users\Alex\anaconda3\lib\site-packages\lasy\utils\mode_decomposition.py:140: ComplexWarning: Casting complex values to real discards the imaginary part
  coeffTest[i] = np.sum(profile * field)
Estimated w0 = 13.20 microns
C:\Users\Alex\anaconda3\lib\site-packages\lasy\utils\mode_decomposition.py:91: ComplexWarning: Casting complex values to real discards the imaginary part
  if math.isnan(coef):
Traceback (most recent call last):

  File "C:\Users\Alex\Desktop\MyWork\sim\Lasy\example_modal_decomposition_data.py", line 59, in <module>
    reconstructedProfile = modeCoeffs[

  File "C:\Users\Alex\anaconda3\lib\site-packages\lasy\profiles\profile.py", line 67, in __rmul__
    return ScaledProfile(self, factor)

  File "C:\Users\Alex\anaconda3\lib\site-packages\lasy\profiles\profile.py", line 128, in __init__
    assert isinstance(factor, (int, float)), "The factor must be a number."

AssertionError: The factor must be a number. 

**how to fix the above error ? May be this example, although not Gerchberg-Saxton algorithm, but may help to understand how to use two intensity distribution images**

Hi @rob-shalloo , thanks a lot for your kind reply and explanation. I am running the example as it is. I downloaded both examples:

(1) example_gerchberg_saxton_algo.py and **(this works without any problem)**
(2) example_modal_decomposition_data.py **(but this throws an error)**

Once again, I downloaded the script file ("example_modal_decomposition_data.py") and run it. The error message was the same:

Estimated w0 = 13.20 microns
Traceback (most recent call last):

File "C:\Users\Alex\Desktop\MyWork\sim\Lasy\example_modal_decomposition_data.py", line 59, in 
reconstructedProfile = modeCoeffs[

File "C:\Users\Alex\anaconda3\lib\site-packages\lasy\profiles\profile.py", line 67, in __rmul__
return ScaledProfile(self, factor)

File "C:\Users\Alex\anaconda3\lib\site-packages\lasy\profiles\profile.py", line 128, in __init__
assert isinstance(factor, (int, float)), "The factor must be a number."

AssertionError: The factor must be a number.

**Although the final goal is to learn how to implement real intensity and phase profile (measured in the experiment) in the simulation, but I would like to go gradually. Foremost, I would like to use / learn both the above examples script and get some idea about the procedure.** 

Also, thanks for the great work of LASY. 

Thanks.
Alex

--- a/lasy/utils/mode_decomposition.py
+++ b/lasy/utils/mode_decomposition.py
@@ -87,7 +87,9 @@ def hermite_gauss_decomposition(laserProfile, n_x_max=12, n_y_max=12, res=1e-6):
     for i in range(n_x_max):
         for j in range(n_y_max):
             HGMode = HermiteGaussianTransverseProfile(w0, i, j)
-            coef = np.real(np.sum(field * HGMode.evaluate(X, Y)) * dx * dy)  # modalDecomposition
+            coef = np.real(
+                np.sum(field * HGMode.evaluate(X, Y)) * dx * dy
+            )  # modalDecomposition
             if math.isnan(coef):
                 coef = 0
             weights[(i, j)] = coef

--- a/lasy/utils/mode_decomposition.py
+++ b/lasy/utils/mode_decomposition.py
@@ -87,7 +87,7 @@ def hermite_gauss_decomposition(laserProfile, n_x_max=12, n_y_max=12, res=1e-6):
     for i in range(n_x_max):
         for j in range(n_y_max):
             HGMode = HermiteGaussianTransverseProfile(w0, i, j)
-            coef = np.sum(field * HGMode.evaluate(X, Y)) * dx * dy  # modalDecomposition
+            coef = np.real(np.sum(field * HGMode.evaluate(X, Y)) * dx * dy)  # modalDecomposition
             if math.isnan(coef):
                 coef = 0
             weights[(i, j)] = coef
@@ -137,7 +137,7 @@ def estimate_best_HG_waist(x, y, field):
         # create a gaussian
         HGMode = HermiteGaussianTransverseProfile(wTest, 0, 0)
         profile = HGMode.evaluate(X, Y)
-        coeffTest[i] = np.sum(profile * field)
+        coeffTest[i] = np.real(np.sum(profile * field))
     w0 = waistTest[np.argmax(coeffTest)]
 
     print("Estimated w0 = %.2f microns" % (w0Est * 1e6))

